### PR TITLE
Basic HW reduced ACPI shutdown/reboot support

### PR DESCRIPTION
**Pull request by rbradford**
- [x] Add reset control I/O port and populate FACP field
- [x] Add shutdown sleep constant (S5) to DSDT
- [x] Add sleep control I/O port and populate FACP field
- [x] Add sleep status I/O port and populate FACP field

### CODE REVIEW

**Comment by rbradford** [comment]
This contains #179 just to allow testing.

**Comment by sboeuf** [comment]
@rbradford After some reading about the `MCFG` table, it looks like we're gonna need it along with the `DSDT` one defining the PCI host bridge. Based on the __PCI Firmware 3.0 specification__:
```
The MCFG table is only used to communicate the base addresses corresponding to the PCI Segment Groups available to the system at boot.  This table directly refers to PCI Segment Groups defined in the system via the _SEG object in the ACPI name space for the applicable host bridge device.
```
One simplification we should follow for now since we're not planning to support multiple segments:
```
For systems containing only a single PCI Segment Group, the default PCI Segment Group number, namely, PCI Segment Group 0, is implied.  In such a case, the default PCI Segment Group need not be represented in the ACPI Name Space (i.e., no _SEG method is required in such a hierarchy).
```

Based on this, my expectation is that we will declare to the guestOS/FW the address of our PCI host bridge to be accessed through MMIO instead of PIO. The problem is that `MCFG` table does not contain important information such as 32bits and 64bits PCI hole associated with this PCI host bridge that usually get defined as part of the DSDT table.
Since we don't have ACPI and the OS uses the entire address space as the PCI hole by default, I'm wondering if same kind of mechanism would apply in the case we would only supply `MCFG`.

I'd say it's worth trying only the `MCFG` since this is clearly the mandatory one as it defines where to find the PCI host bridge, but we might need to set corresponding entry in the DSDT if the guest OS is not happy about not seeing any DSDT entry.

**Comment by rbradford** [comment]
The problem is not the MCFG table (which I have added anyway) but it is instead the expectation from the kernel that it gets the IRQ information from the ACPI tables. You can see this if you trace the effect that "pci=noacpi" has in the kernel source.

**Comment by sboeuf** [comment]
@rbradford is [this](https://github.com/torvalds/linux/blob/master/drivers/acpi/pci_link.c#L582-L597) the code you're referring to?
One question, in case of MSI, ACPI is not used, so why is the Linux kernel expecting to see an IRQ through ACPI tables?

**Comment by rbradford** [comment]
@sboeuf I dug through the code a bit more and I think what it's actually looking for is the _CRS block. This looks like a good source of information about what is needed: https://github.com/torvalds/linux/blob/master/Documentation/PCI/acpi-info.rst

**Comment by sboeuf** [comment]
@rbradford so basically this is the important part we care about:
```
```

**Comment by rbradford** [comment]
@sboeuf We don't need the `_PRT` only the `_CRS` as exemplified by the pci-lite implementation.

**Comment by sboeuf** [comment]
@rbradford great that's good news!

**Comment by sboeuf** [comment]
@rbradford 
I have been able to reproduce the issue you were seeing with the serial IRQ 4 when using ACPI with this PR.
First thing, when things work properly, the IOAPIC advertise that it supports 24 GSIs, and we can see the kernel log:
```
IOAPIC[0]: apic_id 0, version 17, address 0xfec00000, GSI 0-23
```
I have checked and in case of ACPI, our `MADT` table advertise the appropriate address, and also declare the first interrupt as 0. All of this is good, and without ACPI, this causes the OS to start associating MSI interrupt starting from IRQ 24.

Now, the second aspect is that the OS for some reasons decide to ignore the information provided by the HW when we're using ACPI. One thing I found is that if we use the boot option `acpi=noirq`, ACPI is used and everything works well, with MSI IRQs allocated starting at 24.

Looking into the kernel code related to `noirq` option, I found this is basically the option to tell ACPI to ignore PCI IRQ routing. And we know the PCI IRQ routing is done through the `_PRT` of the DSDT but we don't have it since we know we want to use MSI with our devices.
So the first thing that comes to mind is that we could definitely run cloud-hypervisor with the option `acpi=noirq` as a valid configuration IMO.

Problem is, cloud images don't have it, hence I think we should simply define a `_PRT` as part of our DSDT to make the kernel happy. My guess (since I haven't deep dived in the kernel code) is that because the kernel does not find the `_PRT` but expect to find one, it considers that no device is attached to the IOAPIC, and simply ignores the fact that IRQ 0-23 are reserved. I don't see any other reason why the kernel would ignore this information otherwise.

I've started to manually modify the DSDT and try to override it but for some reasons it does not work. I mean not through the rust code, but through the kernel configuration by providing a custom DSDT to my kernel. Here is the simple I'm trying to inject into the DSDT (after the `_CRS`):

```
        Name(_PRT, Package{
            Package{0x0000FFFF, 0, 0, 0},
            Package{0x0000FFFF, 1, 0, 0},
            Package{0x0000FFFF, 2, 0, 0},
            Package{0x0000FFFF, 3, 0, 0},
        })
```
The point is to provide something generic to make the OS happy, but still expecting the MSI to be picked eventually.

**Comment by rbradford** [comment]
@sboeuf So I made some more discoveries on this. If I add a COM1 device then the issues I was having with boot are resolved. However the interrupt for ttyS0 does not appear in /proc/interrupts at all. dmesg reports it has as an IRQ of 0 and the virtio-pci devices start 1:

```
cloud@cloud ~ $ dmesg | grep ttyS0
[    0.000000] Command line:  root=PARTUUID=3cb0e0a5-925d-405e-bc55-edf0cec8f10a console=tty0 console=ttyS0,115200n8 console=hvc0 quiet init=/usr/lib/systemd/systemd-bootchart initcall_debug tsc=reliable no_timer_check noreplace-smp cryptomgr.notests rootfstype=ext4,btrfs,xfs kvm-intel.nested=1 rw
[    0.069606] Kernel command line:  root=PARTUUID=3cb0e0a5-925d-405e-bc55-edf0cec8f10a console=tty0 console=ttyS0,115200n8 console=hvc0 quiet init=/usr/lib/systemd/systemd-bootchart initcall_debug tsc=reliable no_timer_check noreplace-smp cryptomgr.notests rootfstype=ext4,btrfs,xfs kvm-intel.nested=1 rw
[    0.071453] printk: console [ttyS0] enabled
[    0.109042] 00:00: ttyS0 at I/O 0x3f8 (irq = 0, base_baud = 115200) is a U6_16550A
cloud@cloud ~ $ cat /proc/interrupts 
           CPU0       
  1:          0   PCI-MSI 49152-edge      virtio2-config
  2:          7   PCI-MSI 49153-edge      virtio2-input
  3:          0   PCI-MSI 16384-edge      virtio0-config
  4:       1187   PCI-MSI 16385-edge      virtio0-req.0
  5:          0   PCI-MSI 32768-edge      virtio1-config
  6:        128   PCI-MSI 32769-edge      virtio1-req.0
NMI:          0   Non-maskable interrupts
LOC:       7840   Local timer interrupts
SPU:          0   Spurious interrupts
PMI:          0   Performance monitoring interrupts
IWI:          0   IRQ work interrupts
RTR:          0   APIC ICR read retries
RES:          0   Rescheduling interrupts
CAL:          0   Function call interrupts
TLB:          0   TLB shootdowns
ERR:          0
MIS:          0
PIN:          0   Posted-interrupt notification event
NPI:          0   Nested posted-interrupt event
PIW:          0   Posted-interrupt wakeup event
```

Does is matter that the interrupt number is different compared to what were were having before? Everything seems to function correctly?

Although the serial port seems to perform very poorly with this change vs master (e.g. you can visually see the output of dmesg scrolling across the screen.)

**Comment by sboeuf** [comment]
@rbradford 
So after our conversation where I was explaining how removing the IOAPIC definition from the MADT was making things working, I found out that removing the `HW_REDUCED_ACPI` from the FADT flags was also solving the problem. This is kind of weird since I couldn't find in the spec a place where it says it does not support legacy ISA interrupts. This made me think something was tied to the fact that in case of `HW_REDUCED_ACPI`, there are some legacy fallbacks that are implemented by the kernel, based on the standard ACPI spec.

That's why the proper solution is to declare the interrupt source to override to make sure the kernel is aware about it. I think this is the way to tell the kernel how to handle the special 0-15 cases.

I added the patch to this PR: https://github.com/intel/cloud-hypervisor/pull/180/commits/713807bca7b2a4c2a0c9669eb13d87061303dbbd

**Comment by rbradford** [comment]
This patchset uses the same EventFd for shutdown and reboot but sadly that doesn't work out as in some situations, such as the triple-fault each vcpu thread writes to the fd resulting in a different number than expected when running with SMP.

I will update the PR to use separate EventFds for reboot and shutdown. The i8042 and vcpu code only need the reset one, the ACPI support both.

**Comment by rbradford** [comment]
This PR is getting quite large enough already so I opened #221 and #222 to handle some extra items that should be addressed.

**Comment by rbradford** [comment]
I disabled reboot in the least intrusive way for now and created an issue at #223 to capture the work that we need to do before we can turn it back on.

File: src/main.rs [hunk 1]
```diff
@@ -1544,29 +1544,7 @@ mod tests {
 
             thread::sleep(std::time::Duration::new(20, 0));
 
-            // Test that there is no ttyS0
-            aver_eq!(
```
Comment (sboeuf) [inline src/main.rs]: This commit should be removed, and the tests restored.

File: devices/src/legacy/i8042.rs [hunk 2]
```diff
@@ -34,6 +34,7 @@ impl BusDevice for I8042Device {
 
     fn write(&mut self, _base: u64, offset: u64, data: &[u8]) {
```
Comment (sboeuf) [inline devices/src/legacy/i8042.rs]: One typo in the commit message: s/`tiple`/`triple`/

File: arch/src/x86_64/acpi.rs [hunk 3]
```diff
@@ -182,14 +194,21 @@ pub fn create_acpi_tables(
     let mut facp = SDT::new(*b"FACP", 276, 6, *b"CLOUDH", *b"CHFACP  ", 1);
 
     let fadt_flags: u32 = 1 << 20; // HW_REDUCED_ACPI
+    let fadt_flags = fadt_flags | 1 << 10; // RESET_REG_SUP
```
Comment (sboeuf) [inline arch/src/x86_64/acpi.rs]:
```suggestion
    let fadt_flags: u32 = (1 << 10) | (1 << 20);
```

File: devices/src/acpi.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use vmm_sys_util::eventfd::EventFd;
+use BusDevice;
+
+/// A device for handling ACPI shutdown and reboot
+pub struct AcpiDevice {
```
Comment (sboeuf) [inline devices/src/acpi.rs]: I would rename this as `AcpiShutdownDevice` or something similar to make sure we don't think this is some sort of generic ACPI device.

File: devices/src/acpi.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        AcpiDevice {
+            exit_evt,
+            reset_evt,
+        }
+    }
+}
+
+// Same I/O port used for shutdown and reboot
+impl BusDevice for AcpiDevice {
+    fn read(&mut self, _base: u64, _offset: u64, _data: &mut [u8]) {}
```
Comment (sboeuf) [inline devices/src/acpi.rs]: Should we log some warning here since we don't expect this port to be read?

File: src/main.rs [hunk 6]
```diff
@@ -1955,6 +1955,63 @@ mod tests {
         });
     }
 
+    #[test]
```
Comment (sboeuf) [inline src/main.rs]: Typo in the commit message:
s/`with ACPI`/`without ACPI`/

File: devices/src/acpi.rs (hunk 5 above)
Comment (rbradford) [reply devices/src/acpi.rs]: I actually made it return 0 per the spec.

**Comment by rbradford** [comment]
@sboeuf All feedback integrated thank you!

File: devices/src/acpi.rs [hunk 7]
```diff
@@ -0,0 +1,52 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use vmm_sys_util::eventfd::EventFd;
+use BusDevice;
+
```
Comment (sameo) [inline devices/src/acpi.rs]: Here as well: Fine as a first iteration, but the whole ACPI support, including this ACPI device, should be made a build time option.

File: devices/src/acpi.rs (hunk 5 above)
Comment (sboeuf) [reply devices/src/acpi.rs]: Ok!

File: arch/src/x86_64/acpi.rs [hunk 8]
```diff
@@ -93,14 +93,26 @@ pub fn create_acpi_tables(guest_mem: &GuestMemoryMmap, num_cpus: u8) -> GuestAdd
         .expect("Error writing MADT table");
     tables.push(madt_offset.0);
 
+    // MCFG
+    let mut mcfg = SDT::new(*b"MCFG", 60, 1, *b"CLOUDH", *b"CHMCFG  ", 1);
+
+    mcfg.write_u64(44, super::MEM_32BIT_DEVICES_GAP_SIZE);
```
[the remark this answers is not in the document]
Comment (rbradford) [reply arch/src/x86_64/acpi.rs]: I've refactored this code to append a structure instead which should be self-documenting

File: arch/src/x86_64/acpi.rs [hunk 9]
```diff
@@ -27,14 +27,115 @@ struct IOAPIC {
...
+                    0x0000000800000000, // Range Minimum
+                    0x0000000FFFFFFFFF, // Range Maximum
+                    0x0000000000000000, // Translation Offset
+                    0x0000000800000000, // Length
+                    ,, , AddressRangeMemory, TypeStatic)
+            })
+        }
+        */
+
+    let pci_dsdt_data = [
```
[the remark this answers is not in the document]
Comment (rbradford) [reply arch/src/x86_64/acpi.rs]: I documented this in a comment in the function.

**Comment by rbradford** [comment]
@sameo all actionable feedback applied.

**Pull request merged**